### PR TITLE
Added Gateway API HTTPRoute support

### PR DESCRIPTION
**Pull request by networkop**
TODO: 
- [x] Documentation
- [x] tests
- [x] corner cases 
  - [x] more than 1 GW 
  - [x] GW doesn't exist
  - [x] GW namespace is the same/different
  - [x] GW CRDs are not installed

### CODE REVIEW

File: kubernetes.go [hunk 1]
```diff
@@ -106,13 +144,31 @@ func (gw *Gateway) RunKubeController(ctx context.Context) error {
...
+	}
+
+	gw.Controller = newKubeController(ctx, kubeClient, gwAPIClient)
 	go gw.Controller.run()
 
 	return nil
 
 }
 
+func checkGatewayCRDs(ctx context.Context, c *gatewayClient.Clientset) bool {
```
Comment (Kaezon) [inline kubernetes.go:159]: I was trying to figure out the best way to check for the presence of CRDs in the cluster, and this seems effective.
Definitely going to have to add a modified version of this for VirtualServers on my feature branch.

**Comment by Kaezon** [comment]
Everything looks good. My changes are all in more-or-less the same places.
I like the idea of namespacing the optional support test yamls into their own directories, so I'll definitely do that as well.

**Review by Kaezon — APPROVED** [review]

File: gateway_test.go [hunk 2]
```diff
@@ -182,6 +182,20 @@ var tests = []test.Case{
...
 		},
 	},
+	// basic gateway API lookup
+	{
+		Qname: "domain.gw.example.com.", Qtype: dns.TypeA, Rcode: dns.RcodeSuccess,
+		Answer: []dns.RR{
+			test.A("domain.gw.example.com.	60	IN	A	192.0.2.1"),
+		},
+	},
+	// gateway API lookup priority over Ingress
```
[the remark this answers is not in the document]
Comment (networkop) [reply gateway_test.go:192]: yes, hence they are stored in `orderedResources `. It always helps to be explicit about priorities during collisions.

File: kubernetes.go (hunk 1 above)
Comment (networkop) [reply kubernetes.go:159]: great

File: Tiltfile [hunk 3]
```diff
@@ -27,3 +27,8 @@ k8s_yaml('./test/ingress.yaml')
 
 # Metallb
 k8s_yaml('./test/metallb.yaml')
+
+# Gateway API
```
Comment (Kaezon) [inline Tiltfile:31]: Oh, I just realized you might want to add a `k8s_kind()` directive in here for the Gateway objects.
I'm still new to using Tilt, so maybe it isn't necessary, but I did end up adding one for VirtualServers:
`k8s_kind('VirtualServer', api_version='k8s.nginx.org/v1')`

Comment (networkop) [reply Tiltfile:31]: ha, interesting, I haven't seen this before. I did see Tilt complaining about some CRDs but didn't pay attention to it. Thanks, I'll look it up.

**Comment by networkop** [comment]
@Kaezon I want to merge this PR today. Does it look ok to you?

File: kubernetes.go [hunk 4]
```diff
@@ -106,13 +144,31 @@ func (gw *Gateway) RunKubeController(ctx context.Context) error {
...
 	go gw.Controller.run()
 
 	return nil
 
 }
 
+func existGatewayCRDs(ctx context.Context, c *gatewayClient.Clientset) bool {
+
+	_, err := c.GatewayV1alpha2().Gateways("").List(ctx, metav1.ListOptions{})
+	if meta.IsNoMatchError(err) || runtime.IsNotRegisteredError(err) || errors.IsNotFound(err) {
```
Comment (ammmze) [inline kubernetes.go:162]: Should we have added a check on 'forbidden' on here? I just upgraded `k8s_gateway` in my custom coredns build used with the official coredns helm cart and it won't start up because the service account currently doesn't have access to gateways.

I kind of think we should catch the forbidden error and log it with a message saying it doesn't have access (i.e. different from the message here), but probably shouldn't panic when it is forbidden.

```
[WARNING] An external plugin (/go/pkg/mod/github.com/ori-edge/k8s_gateway@v0.2.0/setup.go line 56) is using the deprecated function Normalize. This will be removed in a future versions of CoreDNS. The plugin should be updated to use OriginsFromArgsOrServerBlock or NormalizeExact instead.
[INFO] plugin/k8s_gateway: Building k8s_gateway controller
panic: gateways.gateway.networking.k8s.io is forbidden: User "system:serviceaccount:network:default" cannot list resource "gateways" in API group "gateway.networking.k8s.io" at the cluster scope

goroutine 1 [running]:
github.com/ori-edge/k8s_gateway.existGatewayCRDs({0x214c368, 0xc0001400c8}, 0x1ddc0f4)
    /go/pkg/mod/github.com/ori-edge/k8s_gateway@v0.2.0/kubernetes.go:169 +0x191
github.com/ori-edge/k8s_gateway.newKubeController({0x214c368, 0xc0001400c8}, 0xc00052cf20, 0xc00007bee0)
    /go/pkg/mod/github.com/ori-edge/k8s_gateway@v0.2.0/kubernetes.go:50 +0x11a
github.com/ori-edge/k8s_gateway.(*Gateway).RunKubeController(0xc000176b40, {0x214c368, 0xc0001400c8})
    /go/pkg/mod/github.com/ori-edge/k8s_gateway@v0.2.0/kubernetes.go:152 +0x85
github.com/ori-edge/k8s_gateway.setup(0x1dd8168)
    /go/pkg/mod/github.com/ori-edge/k8s_gateway@v0.2.0/setup.go:29 +0x112
github.com/coredns/caddy.executeDirectives(0xc00021c900, {0x7fff2be57ed8, 0x15}, {0x30d1460, 0x32, 0x1000000004b8066}, {0xc00007e980, 0x1, 0x8}, 0x0)
    /go/pkg/mod/github.com/coredns/caddy@v1.1.1/caddy.go:661 +0x5f6
github.com/coredns/caddy.ValidateAndExecuteDirectives({0x2142d18, 0xc000222e80}, 0xc000222e80, 0x0)
    /go/pkg/mod/github.com/coredns/caddy@v1.1.1/caddy.go:612 +0x3e5
github.com/coredns/caddy.startWithListenerFds({0x2142d18, 0xc000222e80}, 0xc00021c900, 0x0)
    /go/pkg/mod/github.com/coredns/caddy@v1.1.1/caddy.go:515 +0x274
github.com/coredns/caddy.Start({0x2142d18, 0xc000222e80})
    /go/pkg/mod/github.com/coredns/caddy@v1.1.1/caddy.go:472 +0xe5
github.com/coredns/coredns/coremain.Run()
    /go/src/coremain/run.go:63 +0x1cd
main.main()
    /go/src/coredns.go:12 +0x17
```

Comment (networkop) [reply kubernetes.go:162]: yep, totally makes sense, you're right. will try to fix it today
